File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int result = 0;
        int start = 0;
        int end = 0;
        while (end < s.length()) {
            while (set.contains(s.charAt(end))) {
                set.remove(s.charAt(start));
                start++;
            }
            set.add(s.charAt(end));
            end++;
            result = Math.max(result, end - start);
            
        }
        return result;

    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        found = set()
        index = 0
        result = 0
        for i in range(len(s)):
            if s[i] in found:
                while s[i] in found:
                    found.remove(s[index])
                    index += 1
            found.add(s[i])
            result = max(i - index + 1, result)
        return result if result > 0 else len(s)
